libs/h3/src/main/java/org/elasticsearch/h3/HexRing.java

```diff
@@ -581,52 +599,116 @@ public static long[] hexRing(long origin) {
         int idx = 0;
         long previous = -1;
         for (int i = 0; i < 6; i++) {
-            int[] rotations = new int[] { 0 };
-            long[] nextNeighbor = new long[] { 0 };
-            int neighborResult = h3NeighborRotations(origin, DIRECTIONS[i].digit(), rotations, nextNeighbor);
-            if (neighborResult != E_PENTAGON) {
-                // E_PENTAGON is an expected case when trying to traverse off of
+            long neighbor = h3NeighborRotations(origin, DIRECTIONS[i].digit());
+            if (neighbor != -1) {
+                // -1 is an expected case when trying to traverse off of
                 // pentagons.
-                if (neighborResult != E_SUCCESS) {
-                    throw new IllegalArgumentException();
-                }
-                if (previous != nextNeighbor[0]) {
-                    out[idx++] = nextNeighbor[0];
-                    previous = nextNeighbor[0];
+                if (previous != neighbor) {
+                    out[idx++] = neighbor;
+                    previous = neighbor;
                 }
             }
         }
         assert idx == out.length;
         return out;
     }
 
+    /**
+     * Returns whether or not the provided H3Indexes are neighbors.
+     * @param origin The origin H3 index.
+     * @param destination The destination H3 index.
+     * @return true if the indexes are neighbors, false otherwise
+     */
+    public static boolean areNeighbours(long origin, long destination) {
+        // Make sure they're hexagon indexes
+        if (H3Index.H3_get_mode(origin) != Constants.H3_CELL_MODE) {
+            throw new IllegalArgumentException("Invalid cell: " + origin);
+        }
+
+        if (H3Index.H3_get_mode(destination) != Constants.H3_CELL_MODE) {
+            throw new IllegalArgumentException("Invalid cell: " + destination);
+        }
+
+        // Hexagons cannot be neighbors with themselves
+        if (origin == destination) {
+            return false;
+        }
+
+        final int resolution = H3Index.H3_get_resolution(origin);
+        // Only hexagons in the same resolution can be neighbors
+        if (resolution != H3Index.H3_get_resolution(destination)) {
+            return false;
+        }
+
+        // H3 Indexes that share the same parent are very likely to be neighbors
+        // Child 0 is neighbor with all of its parent's 'offspring', the other
+        // children are neighbors with 3 of the 7 children. So a simple comparison
+        // of origin and destination parents and then a lookup table of the children
+        // is a super-cheap way to possibly determine they are neighbors.
```

Comment:
This seems to indicate that the `h3NeighborRotations` method is expensive. I wonder how this will affect our plans to improve the performance of the HexGridTiler by using these methods. Certainly I want to benchmark that,